### PR TITLE
Armed reduce

### PR DESCRIPTION
**Pull request by rcooke-ast**
Echelle reduction is now complete until the extraction step. I'll now work on getting the extraction to work on multiple "slits" (i.e. orders) which should also be useful for the multislit reductions. Here are some specifics that I've added with this PR:

* New settings for object tracing
* Object tracing for echelle data (including PCA of object trace)
* Added "pixel" wavelength solution (when the user does not want to perform a wavelength solution)

### CODE REVIEW

File: pypit/armasters.py [hunk 1]
```diff
@@ -163,8 +163,11 @@ def save_masters(slf, det, setup):
                            frametype='wave')
         # Wavelength fit
         gddict = ltu.jsonify(slf._wvcalib[det-1])
-        json_file=master_name('wave_calib', setup)
-        ltu.savejson(json_file, gddict, easy_to_read=True, overwrite=True)
+        json_file = master_name('wave_calib', setup)
+        if gddict is not None:
+            ltu.savejson(json_file, gddict, easy_to_read=True, overwrite=True)
+        else:
```
Comment (profxj) [inline pypit/armasters.py]: Is it that it hasn't been saved or that it was never
generated?

Comment (rcooke-ast) [inline pypit/armasters.py]: When the user requests a "pixel" wavelength solution (i.e. specifically requests to not perform a wavelength solution) then gddict is not created.

File: pypit/arproc.py [hunk 2]
```diff
@@ -857,46 +882,96 @@ def reduce_echelle(slf, sciframe, scidx, fitsdict, det, standard=False):
         slf._bgframe[det - 1] = bgframe
     # Obtain a first estimate of the object trace then
     # fit the traces and perform a PCA for the refinements
-    trccoeff = np.zeros((settings.argflag['trace']['object']['order'], nord))
+    trccoeff = np.zeros((settings.argflag['trace']['object']['order']+1, nord))
     trcxfit = np.arange(nspec)
     for o in range(nord):
         trace, error = artrace.trace_weighted(sciframe-bgframe, slf._lordloc[det-1][:, o], slf._rordloc[det-1][:, o],
                                               mask=slf._scimask[det-1], wght="flux")
+        trace -= slf._lordloc[det-1][:, o]
```
Comment (profxj) [inline pypit/arproc.py]: Can you add a comment here on the reasoning for 
this manipulation?  Thanks.

File: pypit/arproc.py [hunk 3]
```diff
@@ -857,46 +882,96 @@ def reduce_echelle(slf, sciframe, scidx, fitsdict, det, standard=False):
...
-    msgs.info("Performing a PCA on the object trace")
-    ofit = settings.argflag['trace']['object']['pca']['params']
-    lnpc = len(ofit) - 1
-    msgs.work("May need to do a check here to make sure ofit is reasonable")
-    xcen = trcxfit[:, np.newaxis].repeat(nord, axis=1)
-    trccen = arutils.func_val(trccoeff, trcxfit, settings.argflag['trace']['object']['function'],
-                              minv=0.0, maxv=nspec-1.0)
-    fitted, outpar = arpca.basis(xcen, trccen, trccoeff, lnpc, ofit, skipx0=False,
-                                 function=settings.argflag['trace']['object']['function'])
+    refine = 0.0
```
Comment (profxj) [inline pypit/arproc.py]: Is all of this in the reduce_echelle() method?

Seems like it might be best to modularize more.

File: pypit/arproc.py [hunk 4]
```diff
@@ -857,46 +882,96 @@ def reduce_echelle(slf, sciframe, scidx, fitsdict, det, standard=False):
...
+    trobjr = trccen + trobjr
+    # Convert trccen to the actual trace locations
+    trccen *= (slf._rordloc[det - 1] - slf._lordloc[det - 1])
+    trccen += slf._lordloc[det - 1]
+    trobjl *= (slf._rordloc[det - 1] - slf._lordloc[det - 1])
+    trobjl += slf._lordloc[det - 1]
+    trobjr *= (slf._rordloc[det - 1] - slf._lordloc[det - 1])
+    trobjr += slf._lordloc[det - 1]
+
+    # Generate an image of pixel weights for each object
```
Comment (profxj) [inline pypit/arproc.py]: Maybe specify what the weight factor ranges from.
Are they mainly 0's and 1's ?

File: pypit/artrace.py [hunk 5]
```diff
@@ -294,42 +294,147 @@ def expand_slits(slf, mstrace, det, ordcen, extord):
     return lordloc, rordloc
 
 
-def trace_object(slf, det, sciframe, varframe, crmask, trim=2.0,
+def trace_objbg_image(slf, det, sciframe, slitn, objreg, bgreg, trim=2, triml=None, trimr=None):
+    """ Creates an image with weights corresponding to object or background pixels
```
Comment (profxj) [inline pypit/artrace.py]: Are the weights simply either 0 or 1 ?

File: pypit/artrace.py [hunk 6]
```diff
@@ -294,42 +294,147 @@ def expand_slits(slf, mstrace, det, ordcen, extord):
...
+    if trimr is None:
+        trimr = trim
+    npix = int(slf._pixwid[det-1][slitn] - triml - trimr)
+    # Make an image of pixel weights for each object
+    xint = np.linspace(0.0, 1.0, sciframe.shape[0])
+    yint = np.linspace(0.0, 1.0, npix)
+    yint = np.append(-yint[1], np.append(yint, 2.0-yint[-2]))
+    msgs.info("Creating an image weighted by object pixels")
+    rec_obj_img = np.zeros((sciframe.shape[0], sciframe.shape[1], nobj))
+    for o in range(nobj):
```
Comment (profxj) [inline pypit/artrace.py]: As you are looping on object,  what kind of values
result when objects overlap?

File: pypit/artrace.py [hunk 7]
```diff
@@ -294,42 +294,147 @@ def expand_slits(slf, mstrace, det, ordcen, extord):
...
+        idxarr = np.zeros(sciframe.shape).flatten()
+        idxarr[wf] = 1
+        idxarr = idxarr.reshape(sciframe.shape)
+        rec_img = np.zeros_like(sciframe)
+        rec_img[np.where(idxarr == 1)] = rec_obj_arr
+        rec_obj_img[:, :, o] = rec_img.copy()
+    # Make an image of pixel weights for the background region of each object
+    msgs.info("Creating an image weighted by background pixels")
+    rec_bg_img = np.zeros((sciframe.shape[0], sciframe.shape[1], nobj))
+    for o in range(nobj):
```
Comment (profxj) [inline pypit/artrace.py]: Same question as above.

File: pypit/artrace.py [hunk 8]
```diff
@@ -294,42 +294,147 @@ def expand_slits(slf, mstrace, det, ordcen, extord):
...
 
     Returns
     -------
-
+    tracedict : dict
+      A dictionary containing the object trace information
     """
     from pypit import arcytrace
     from pypit import arcyutils
+    # Find the trace of each object
```
Comment (profxj) [inline pypit/artrace.py]: Should you add some of these parameters to the tracedict for
posterity?

File: pypit/armasters.py (hunk 1 above)
Comment (profxj) [inline pypit/armasters.py]: Ok, got it.

File: pypit/arparse.py [hunk 9]
```diff
@@ -1690,6 +1701,31 @@ def trace_dispersion_direction(self, v):
...
 
+    def trace_object_function(self, v):
+        """ What function should be used to trace the object in each order?
+
+        Parameters
+        ----------
+        v : str
+          value of the keyword argument given by the name of this function
+        """
+        allowed = ['polynomial', 'legendre', 'chebyshev']
```
Comment (rcooke-ast) [inline pypit/arparse.py]: There are quite a few instances of this throughout arparse.py...

I had originally intended to write a script that would automatically generate a .rst file, containing the doc string for each argument combined with the allowed values. So, if a new parameter is added, we wouldn't have to also update the docstring. I don't think the typical user will go to the docstring to find the allowed values, but is more more likely to go to the documentation. Do you have a preference here?

Comment (profxj) [inline pypit/arparse.py]: I like your suggestion for parsing into .rst files.  
Let's aim for that.

**Comment by profxj** [comment]
I see 2 lingering comments on artrace but am
happy for you to merge.

**Comment by rcooke-ast** [comment]
Sorry - I just realized that I had written responses but not "accepted" the review responses... just added.

File: pypit/artrace.py (hunk 6 above)
Comment (profxj) [inline pypit/artrace.py]: Ok, got it.

Close pairs of sources are a challenge (and one we will
deal with soon).

**Comment by profxj** [comment]
Merging.

File: pypit/arproc.py (hunk 3 above)
Comment (profxj) [inline pypit/arproc.py]: I was mainly remarking that this function was getting quite long
so any effort to break it up would be good.

**Pull request merged**